### PR TITLE
All LearningModelSessions created from a common LearningModelDevice should share the same thread pool

### PR DESCRIPTION
**Pull request by smk2007**
All LearningModelSessions created from a common LearningModelDevice should share the same thread pool

Issue: When WinML is used in a calling pattern than ping pongs between 2 different models (A, and B), each session is created with a unique thread pool. These pools are configured to use a number of threads equivalent to std::thread :hardware_concurrency.

Since both session utilize thread spinning for performance, even when model A is not being evaluated, its threads are spinning and wasting cpu resources. This causes model B to slow down.

Fix: LearningModelSessions should share common device resources when they share the same LearningModelDevice. The fix shares a common thread pool across LearningModelSessions, and enables sequential execution of different models without creating unnecessary threads that drain cpu time. 

Tests have indicated that 10-20% improvement in performance is achievable by caching the device.

### CODE REVIEW

File: winml/lib/Api/LearningModelSession.cpp [hunk 1]
```diff
@@ -128,23 +129,29 @@
...
-    // If session options aren't provided, force the thread pool size to be maxxed out
-    // to ensure that WinML always runs the fastest.
-    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpNumThreadsOverride(std::thread::hardware_concurrency()));
+    winrt::com_ptr<_winml::IThreading> thread_pool = nullptr;
+    WINML_THROW_IF_FAILED(device_impl->GetThreadPool(thread_pool.put()));
+    if (thread_pool == nullptr)
+    {
+      WINML_THROW_IF_FAILED(engine_factory_->CreateThreadPool(allow_spinning, num_intra_op_threads, thread_pool.put()));
+      WINML_THROW_IF_FAILED(device_impl->CacheThreadPool(thread_pool.get()));
+    }
```
Comment (smk2007) [inline winml/lib/Api/LearningModelSession.cpp:153]: The weirdness here is so that the shared thread pool on the device can be used to honor the session options (spinning, and intra op threads counts).

Ideally these would be optoins exposed on the device and not randomly picked up from the first session created with the device.

In this scheme the session options are always honored, but the shared thread pool on the device is configured based on the first session created.

The end user should never end up with behavior that they dont like, except possibly additional threads created when the internal device's thread pool was configured in an unexpected way (1st session's properties).

Since this behavior is undefined anyway, im okay with keeping this weird, as it provides some mechanism for exposing the spinning and num threads options to the device's shared thread pool (albeit unexpectedly).

A future API change is required to make the device's spinning and thread count properties available. This will allow the construction of the thread pool inside LearnignModelDevice and not as a side effect of Session creation.

File: winml/test/scenario/cppwinrt/scenariotestscppwinrt.cpp [hunk 2]
```diff
@@ -136,7 +136,10 @@ static void Scenario1LoadBindEvalDefault() {
   std::wstring filePath = FileHelpers::GetModulePath() + L"model.onnx";
   LearningModel model = LearningModel::LoadFromFilePath(filePath);
   // create a session on the default device
-  LearningModelSession session(model, LearningModelDevice(LearningModelDeviceKind::Default));
+  auto device = LearningModelDevice(LearningModelDeviceKind::Default);
+  LearningModelSession session(model, device);
+  LearningModelSession session2(model, device);
+  LearningModelSession session3(model, device);
```
Comment (linnealovespie) [inline winml/test/scenario/cppwinrt/scenariotestscppwinrt.cpp:142]: Is this just to show that you can create multiple sessions with the same device, or are sessions 2 and 3 used for something else?

File: onnxruntime/core/session/inference_session.h [hunk 3]
```diff
@@ -132,8 +132,12 @@ class InferenceSession {
   explicit InferenceSession(const SessionOptions& session_options,
                             const Environment& session_env);
 
-#if !defined(ORT_MINIMAL_BUILD)
+  explicit InferenceSession(const SessionOptions& session_options,
+                            const Environment& session_env,
+                            onnxruntime::concurrency::ThreadPool* intra_op_thread_pool,
```
Comment (snnn) [inline onnxruntime/core/session/inference_session.h:137]: OrtEnv already has a global thread pool. Is this change really needed?

Comment (smk2007) [reply onnxruntime/core/session/inference_session.h:137]: The global thread pool cannot be used in winml, as there is no global environment object exposed. We have a device object with is the natural place to own these settings.

Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.h:137]: 1. Can these parameters be named as external_?
2. Also, we need documentation to justify the addition of this overload. Please state explicitly that it was added for WinML only.

File: winml/test/scenario/cppwinrt/scenariotestscppwinrt.cpp (hunk 2 above)
Comment (smk2007) [reply winml/test/scenario/cppwinrt/scenariotestscppwinrt.cpp:142]: just to create the sessions - and the thread pools

File: onnxruntime/core/session/inference_session.h (hunk 3 above)
Comment (smk2007) [reply onnxruntime/core/session/inference_session.h:137]: Sure

**Review by linnealovespie — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]
